webserver/src/main.rs

```diff
@@ -0,0 +1,417 @@
+                .unwrap_or(("-".into(), "-".into()));
+            let red_team = format!("{}, {}", game.player_red_a, game.player_red_b);
+            let blue_team = format!("{}, {}", game.player_blue_a, game.player_blue_b);
+            let (winners, losers, drawers) = match game.result.as_str() {
+                "DRAW" => (
+                    "".to_string(),
+                    "".to_string(),
+                    format!("{}, {}", red_team, blue_team),
+                ),
+                "VICTORY_RED" => (red_team, blue_team, "".to_string()),
+                "VICTORY_BLUE" => (blue_team, red_team, "".to_string()),
+                _ => ("".to_string(), "".to_string(), "".to_string()),
+            };
+            rsx! {<tr>
+                <td>{start_date}</td>
+                <td>{start_time}</td>
+                <td>{winners}</td>
+                <td>{losers}</td>
+                <td>{drawers}</td>
+                <td><a href={format!("/dyn/pgn/{}", rowid.id)}>{"pgn💾"}</a></td>
+            </tr>}
+        })
+        .collect::<Vec<_>>();
+
+    let h: String = html! {
+        <html>
+            <style>
+                {STYLESHEET}
+            </style>
+        <head>
+        </head>
+        <body>
+          <table>
+            <tr>
+                <th>{"Date"}</th>
+                <th>{"Time"}</th>
+                <th>{"Winners"}</th>
+                <th>{"Losers"}</th>
+                <th>{"Drawers"}</th>
+                <th>{"Pgn"}</th>
+            </tr>
+            {table_body}
+          </table>
+        </body>
+        </html>
+    };
+    let mut resp = Response::new(StatusCode::Ok);
+    resp.set_content_type(Mime::from("text/html; charset=UTF-8"));
+    resp.set_body(h);
+    Ok(resp)
+}
+
+async fn pgn<D: DatabaseReader>(req: Request<D>) -> tide::Result {
+    let rowid = req.param("rowid")?.parse()?;
+    let p = req.state().pgn(RowId { id: rowid })?;
+    let mut resp = Response::new(StatusCode::Ok);
+    resp.insert_header(
+        "Content-Disposition",
+        format!("attachment; filename=\"game{rowid}.pgn\""),
+    );
+    resp.set_body(p);
+    Ok(resp)
+}
+
+#[derive(Debug, Clone, Copy, Default)]
+struct RawStats {
+    draws: usize,
+    wins: usize,
+    losses: usize,
+}
+
+#[derive(Debug, Clone, Default)]
+struct FinalStats {
+    name: String,
+    games: usize,
+    draws: usize,
+    wins: usize,
+    losses: usize,
+    points: f64,
+    pointrate: f64,
+}
+
+async fn stats_with_duration<D: DatabaseReader>(req: Request<D>) -> tide::Result {
+    let duration_str = req.param("duration")?;
+    let duration = humantime::parse_duration(duration_str)?;
+    stats(req, Some(duration.try_into()?)).await
+}
+
+async fn stats<D: DatabaseReader>(
+    req: Request<D>,
+    lookback: Option<time::Duration>,
+) -> tide::Result {
+    let now = OffsetDateTime::now_utc();
+    let range_start = match lookback {
+        None => OffsetDateTime::UNIX_EPOCH,
+        Some(d) => now - d,
```

Comment:
IIUC, this would crash if `now < d` (which is possible, because clocks are not guaranteed to be monotonous).
I suggest replacing with [saturating_sub](https://doc.rust-lang.org/std/time/struct.Duration.html#method.saturating_sub)